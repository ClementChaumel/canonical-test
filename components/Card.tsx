import { Post } from '../@types/BlogPostType'
import { getData } from '../utils/embedded-wp-utils'
import { formatDate } from '../utils/date-utils'

interface Card {
  post: Post
}

export const Card: React.FC<Card> = ({ post }) => {
  const {
    title: { rendered: postTitle },
    featured_media: image,
    link,
    date,
    _embedded: embedded,
  } = post

  const { author, category, firstTag: tag, topic } = getData(embedded)

  return (
    <div>
      <h3>{topic ? topic : tag ? tag : 'Canonical News'}</h3>
      <img src={image} alt={`${postTitle} illustration`}></img>
      <a href={link} target="_blank" rel="noopener noreferrer">
        <p>{postTitle}</p>
      </a>
      <p>
        By{' '}
        <a href={author.link} target="_blank" rel="noopener noreferrer">
          {author.name}
        </a>{' '}
        on {formatDate(date)}
      </p>
      <h4>{category}</h4>
    </div>
  )
}

export default Card
