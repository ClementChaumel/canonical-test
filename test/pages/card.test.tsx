import React from 'react'
import { render } from '@testing-library/react'
import { Card } from '../../components/Card'
import {
  mockPost,
  mockPostNoTopic,
  mockPostNoTopicNoTag,
} from '../__mocks__/postMock'

describe('Card component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Card post={mockPost} />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly', () => {
    const { getByText, queryByText, getByAltText } = render(
      <Card post={mockPost} />,
      {}
    )
    getByText('People and culture')
    expect(getByAltText('Post title illustration')).toHaveProperty(
      'src',
      'https://assets.ubuntu.com/v1/2eec3fd9-logo.svg'
    )
    expect(getByText('Post title').parentElement).toHaveProperty(
      'href',
      'https://canonical.com/'
    )
    expect(getByText('Billy Bob')).toHaveProperty(
      'href',
      'https://canonical.com/billy-bob'
    )
    expect(queryByText('24 August 2020')).not.toBeNull
    getByText('Articles')
  })

  it('displays the first tag if there is no topic', () => {
    const { getByText } = render(<Card post={mockPostNoTopic} />, {})

    getByText('careers')
  })

  it('displays the generic message if neither topic nor tag is present', () => {
    const { getByText } = render(<Card post={mockPostNoTopicNoTag} />, {})
    getByText('Canonical News')
  })
})
