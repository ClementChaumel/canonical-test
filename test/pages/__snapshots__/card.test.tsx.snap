// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      People and culture
    </h3>
    <img
      alt="Post title illustration"
      src="https://assets.ubuntu.com/v1/2eec3fd9-logo.svg"
    />
    <a
      href="https://canonical.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <p>
        Post title
      </p>
    </a>
    <p>
      By 
      <a
        href="https://canonical.com/billy-bob"
        rel="noopener noreferrer"
        target="_blank"
      >
        Billy Bob
      </a>
       on 24 August 2020
    </p>
    <h4>
      Articles
    </h4>
  </div>
</DocumentFragment>
`;
